dev/Installer/src/install.cpp

```diff
@@ -0,0 +1,172 @@
+#include "pch.h"
+#include "packages.h"
+#include "install.h"
+
+using namespace winrt;
+using namespace Windows::ApplicationModel;
+using namespace Windows::Foundation;
+using namespace Windows::Management::Deployment;
+using namespace Windows::System;
+
+winrt::hresult add_package(Uri packageUri)
+{
+    PackageManager packageManager;
+    auto deploymentOperation{ packageManager.AddPackageAsync(packageUri, nullptr, DeploymentOptions::None) };
+    deploymentOperation.get();
+    if (deploymentOperation.Status() != AsyncStatus::Completed)
+    {
+        return deploymentOperation.ErrorCode();
+    }
+
+    return S_OK;
+}
+
+winrt::hresult provision_package(std::wstring packageFamilyName)
+{
+    PackageManager packageManager;
+    auto deploymentOperation{ packageManager.ProvisionPackageForAllUsersAsync(packageFamilyName.c_str()) };
+    deploymentOperation.get();
+
+    if (deploymentOperation.Status() != AsyncStatus::Completed)
+    {
+        return deploymentOperation.ErrorCode();
+    }
+
+    return S_OK;
+}
+
+bool is_architecture_applicable(ProcessorArchitecture arch)
+{
+    SYSTEM_INFO systemInfo;
+    GetNativeSystemInfo(&systemInfo);
+    ProcessorArchitecture systemArch = static_cast<ProcessorArchitecture>(systemInfo.wProcessorArchitecture);
+
+    if (arch == systemArch)
+    {
+        return true;
+    }
+
+    // On x64 systems, x86 architecture is also applicable
+    if (systemArch == ProcessorArchitecture::X64 && arch == ProcessorArchitecture::X86)
+    {
+        return true;
+    }
+
+    return false;
+}
+
+wil::com_ptr<IStream> get_resource_stream(std::wstring resourceName, std::wstring resourceType)
+{
+    HMODULE const hmodule = GetModuleHandle(NULL);
+    HRSRC hrsrc = ::FindResource(hmodule, resourceName.c_str(), resourceType.c_str());
+    THROW_LAST_ERROR_IF_NULL(hrsrc);
+    HGLOBAL hresource = LoadResource(hmodule, hrsrc);
+    THROW_LAST_ERROR_IF_NULL(hresource);
+    const BYTE* data = reinterpret_cast<BYTE*>(::LockResource(hresource));
+    THROW_LAST_ERROR_IF_NULL(data);
+    DWORD size = ::SizeofResource(hmodule, hrsrc);
+    wil::com_ptr<IStream> stream = ::SHCreateMemStream(data, size);
```

Comment:
Sadly, this causes a leak - `SHCreateMemStream` returns a stream with a +1 refcount, and the ctor here bumps it with another ref.  Maybe write a function like this:

```c++
wil::com_ptr<IStream> create_mem_stream(const void* data, size_t size)
{
    wil::com_ptr<IStream> retval;
    retval.attach(::SHCreateMemStream(data, size));
    return retval;
}
```

Reply:
I note that this "trap for the unwary" is not noted on the documentation for SHCreateMemStream.